### ISSUE TITLE
Hermes is extremely slow (compared to Chakra) in Debug builds

### ISSUE THREAD

**Issue by asklar** [microsoft/react-native-windows#8024]
- Open `packages\playground\windows\playground-win32.sln`
- Run it, File->Settings->disable web debugging
- File->Open -> RNTester

By default playground-win32 will use Hermes. Using hermes it takes about 20-30s between being clicking and anything happening. Disabling Hermes (and using Chakra) in `packages\playground\windows\ExperimentalFeatures.props`, rebuilding and trying the steps above, results in the app functioning normally.

**Comment by asklar** [comment]
CC @mganandraj @acoates-ms

**Comment by mganandraj** [comment]
@alex very likely the app is loading plain Javascript here, not Hermes bytecode. There is a build task to compile js to byte bytecode when building release flavor though.

Performance of Hermes on debug build is a real concern though. As Hermes don't have a safe abi, we can't use release binaries on debug app, unlike Chakra.

MSVC debug builds are notorious for being extremely slow compared to release builds, and the biggest contributor to the slowness is said to be strong iterator checks. One option is to try disabling iterator debugging, but it's  a global solution level setting. I've plans to run experiments with such settings  .

One good news is that the debug builds are made to run faster with vs22 tool chains.

**Comment by asklar** [comment]
Yes this was in Debug. Haven't tried Release since playground isn't set up for it with RNTester afaik. 
@chiaramooney is gallery set up to use Hermes or is it using Chakra still?

**Comment by chiaramooney** [comment]
@asklar Right now it isn't. But the copy of gallery that's currently previewing 0.65 has the option set up in ExperimentalFeatures; it's just disabled right now.

**Comment by asklar** [comment]
there is a macro `_ITERATOR_DEBUG_LEVEL` that can be used to disable checked iterators

**Comment by acoates-ms** [comment]
yup, the problem is that the api boundary between RNW and the JS engine isn't ABI safe yet  (pending the NAPI work).  So for now, to change that level in Hermes we'll have to change it in RNW too. -- Which I think it what @mganandraj was going to look into -- at least to see if that makes things significantly faster in debug.

**Comment by mganandraj** [comment]
As we don't have abi safe boundary, the iterator level needs to be changed globally.. I haven't got a chance to try it yet.. Will share the details once I have it.

**Comment by mganandraj** [comment]
The debug performance should be good in main now. We have achieved it by using a modified release flavored hermes binary in debug builds of android. In the modified release flavored hermes binary, we set the STL iterator debug level to match the debug flavor, and doing some wrapping of STL containers.

After further investigation and chat with the fb hermes team, we have identified an alternate and more elegant solution (essentially to use hermes debug builds with C++ optimisations turned on) which we will make soon, but it won't impact the final consumption experience in RNW.
